zokrates_core/src/static_analysis/propagation.rs

```diff
@@ -912,147 +940,150 @@ impl<'ast, 'a, T: Field> ResultFolder<'ast, T> for Propagator<'ast, 'a, T> {
-                (e1, e2) => FieldElementExpression::Mult(box e1, box e2),
+                (e1, e2) => Ok(FieldElementExpression::Mult(box e1, box e2)),
             },
             FieldElementExpression::Div(box e1, box e2) => match (
                 self.fold_field_expression(e1)?,
                 self.fold_field_expression(e2)?,
             ) {
                 (FieldElementExpression::Number(n1), FieldElementExpression::Number(n2)) => {
-                    FieldElementExpression::Number(n1 / n2)
+                    Ok(FieldElementExpression::Number(n1 / n2))
                 }
-                (e1, e2) => FieldElementExpression::Div(box e1, box e2),
+                (e1, e2) => Ok(FieldElementExpression::Div(box e1, box e2)),
             },
             FieldElementExpression::Neg(box e) => match self.fold_field_expression(e)? {
-                FieldElementExpression::Number(n) => FieldElementExpression::Number(T::zero() - n),
-                e => FieldElementExpression::Neg(box e),
+                FieldElementExpression::Number(n) => {
+                    Ok(FieldElementExpression::Number(T::zero() - n))
+                }
+                e => Ok(FieldElementExpression::Neg(box e)),
             },
             FieldElementExpression::Pos(box e) => match self.fold_field_expression(e)? {
-                FieldElementExpression::Number(n) => FieldElementExpression::Number(n),
-                e => FieldElementExpression::Pos(box e),
+                FieldElementExpression::Number(n) => Ok(FieldElementExpression::Number(n)),
+                e => Ok(FieldElementExpression::Pos(box e)),
             },
             FieldElementExpression::Pow(box e1, box e2) => {
                 let e1 = self.fold_field_expression(e1)?;
                 let e2 = self.fold_uint_expression(e2)?;
                 match (e1, e2.into_inner()) {
                     (_, UExpressionInner::Value(ref n2)) if *n2 == 0 => {
-                        FieldElementExpression::Number(T::from(1))
+                        Ok(FieldElementExpression::Number(T::from(1)))
                     }
                     (FieldElementExpression::Number(n1), UExpressionInner::Value(n2)) => {
-                        FieldElementExpression::Number(n1.pow(n2 as usize))
+                        Ok(FieldElementExpression::Number(n1.pow(n2 as usize)))
                     }
-                    (e1, UExpressionInner::Value(n2)) => FieldElementExpression::Pow(
+                    (e1, UExpressionInner::Value(n2)) => Ok(FieldElementExpression::Pow(
                         box e1,
                         box UExpressionInner::Value(n2).annotate(UBitwidth::B32),
-                    ),
-                    (_, e2) => unreachable!(format!(
-                        "non-constant exponent {} detected during static analysis",
-                        e2.annotate(UBitwidth::B32)
                     )),
+                    (_, e2) => Err(Error::NonConstantExponent(format!(
```

Comment:
I'm surprised clippy doesnt suggest using `to_string` instead of `format!`

Reply:
that is surprising :|, will refactor this to use `to_string`